fix std.fs.Dir.makePath silent failure

amiralawi opened the pull request:
std.fs.dir.makePath silently fails if one of the items in the path already exists. For example:

cwd.makePath("foo/bar/baz")

Silently failing is OK if "bar" is already a directory or a symlink that resolves to a directory - this is the intended use of makePath (like mkdir -p). But if bar is a file then the subdirectory baz cannot be created - the end result is that makePath doesn't do anything and never creates the directory baz which should be a detectable error.

The existing code had a TODO comment that did not specifically cover this error, but the solution for this silent failure also accomplishes the TODO task - the code now stats "foo" and returns an appropriate error if necessary. The new code also handles potential race condition if "bar" is deleted/permissions changed/etc in between the initial makeDir and statFile calls.

squeek502 commented:
Would be good to add a test case to https://github.com/ziglang/zig/blob/master/lib/std/fs/test.zig

squeek502 commented on lib/std/fs.zig:
```diff
@@ -1469,9 +1469,11 @@ pub const Dir = struct {
                     // this is important because otherwise a dangling symlink
                     // could cause an infinite loop
+                    var fstat = try self.statFile(sub_path);
+                    if(fstat.kind != .directory){ return error.PathIsNotDir; }
```
```suggestion
                    if (fstat.kind != .directory) return error.NotDir;
```

to match the named used in other fs error sets.

squeek502 commented on lib/std/fs.zig:
```diff
@@ -1469,9 +1469,11 @@ pub const Dir = struct {
+                    // stat the file and return an error if it's not a directory
                     // this is important because otherwise a dangling symlink
                     // could cause an infinite loop
+                    var fstat = try self.statFile(sub_path);
```
This will run into https://github.com/ziglang/zig/issues/16738 on Windows and fail with `error.IsDir` if the path is a directory.

amiralawi commented:
I updated the solution per your above comments and added a test to std/fs/test.zig.  Tests on linux & windows OK - I don't have the capability to test on MacOS but don't anticipate that this should be significantly different.

squeek502 commented on lib/std/fs/test.zig:
```diff
@@ -907,6 +907,25 @@ test "makePath in a directory that no longer exists" {
     try testing.expectError(error.FileNotFound, tmp.dir.makePath("sub-path"));
 }
 
+test "makePath but sub_path contains pre-existing file" {
+    // makePath tmp/foo/bar/baz, but bar is a file
+    var buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
+    var fba = std.heap.FixedBufferAllocator.init(&buf);
+    var buf_alloc = fba.allocator();
+
+    var tmp = tmpDir(.{});
+    defer tmp.cleanup();
+    
+    try tmp.dir.makeDir("foo");
+    var foo = try tmp.dir.openDir("foo", .{});
+    defer foo.close();
+    var bar = try foo.createFile("bar", .{});
+    defer bar.close();
+    
+    var sub_path = try std.fs.path.join(buf_alloc, &[_][]const u8{"foo", "bar", "baz"});
+    try testing.expectError(error.NotDir, tmp.dir.makePath(sub_path));
+}
```
Just to simplify things a bit:

```suggestion
test "makePath but sub_path component is a pre-existing file" {
    var tmp = tmpDir(.{});
    defer tmp.cleanup();

    try tmp.dir.makeDir("foo");
    try tmp.dir.writeFile("foo/bar", "");
    
    try testing.expectError(error.NotDir, tmp.dir.makePath("foo/bar/baz"));
}
```

amiralawi replied:
I like the simplification, but is this guaranteed to work in all instances regardless of what the native path separator is?

squeek502 replied:
The only target it would fail on is UEFI (which only supports `\` AFAIK), but there are other tests in `fs/test.zig` that use `/` as the path separator so it should be fine for now:

https://github.com/ziglang/zig/blob/ddf5859c22527c6bf5d8bb13310db996fcc58874/lib/std/fs/test.zig#L454

squeek502 replied:
If you want to use the native path separator, though, this would also work:

```zig
test "makePath but sub_path component is a pre-existing file" {
    var tmp = tmpDir(.{});
    defer tmp.cleanup();
    try tmp.dir.makeDir("foo");
    try tmp.dir.writeFile("foo" ++ fs.path.sep_str ++ "bar", "");
    
    try testing.expectError(error.NotDir, tmp.dir.makePath("foo" ++ fs.path.sep_str ++ "bar" ++ fs.path.sep_str ++ "baz"));
}

squeek502 commented:
Made a PR that will fix the WASI compilation error you're hitting: https://github.com/ziglang/zig/pull/17029

squeek502 commented on lib/std/fs/test.zig:
```diff
@@ -907,6 +907,16 @@ test "makePath in a directory that no longer exists" {
     try testing.expectError(error.FileNotFound, tmp.dir.makePath("sub-path"));
 }
 
+test "makePath but sub_path contains pre-existing file" {
```
If you rebase on master, the WASI failure should be fixed (https://github.com/ziglang/zig/pull/17029), but it'll start hitting https://github.com/ziglang/zig/issues/17034 so you'll need to add this conditional to the test case:

```suggestion
test "makePath but sub_path contains pre-existing file" {
    // TODO: Re-enable once https://github.com/ziglang/zig/issues/17034 is solved
    if (builtin.os.tag == .linux and builtin.link_libc and builtin.abi == .gnu) return error.SkipZigTest;

```

andrewrk commented:
Thank you @amiralawi, and @squeek502 for helping get this PR to completion.
